src/coreclr/src/vm/weakreferencenative.cpp

```diff
@@ -495,7 +496,7 @@ void FinalizeWeakReference(Object * obj)
 
     WEAKREFERENCEREF pThis((WeakReferenceObject *)(obj));
 
-    OBJECTHANDLE handle = AcquireWeakHandleSpinLock(pThis);
+    OBJECTHANDLE handle = ThreadSuspend::SysIsSuspended() ? pThis->m_Handle.LoadWithoutBarrier() : AcquireWeakHandleSpinLock(pThis);
```

Comment:
What is the relationship of background GC to thread suspension. In particular, I am interested in whether or not background GC can run while the runtime is suspended for a non-GC purpose, such as a profiler call, or tiered compilation operation. In that case, there may be some risk of doing this load, then the runtime resuming while the HandleFetchType is happening, and then *something bad* happening.

Also, you need a comment around the ReleaseWeakHandleSpinLock call to indicate why its safe to call that function without having acquired a spinlock.

Reply:
> I am interested in whether or not background GC can run while the runtime is suspended for a non-GC purpose, such as a profiler call, or tiered compilation operation.

My understanding is no, background GC cannot run while the runtime is suspended.

The background GC thread is a managed thread - if someone tries the suspend the runtime, the background GC thread will be suspended as well. The suspension will not be complete until the background GC reach a spot where `allow_fgc()` is called so that its gives up and allow it.

Reply:
My reading of the source code indicates that the logic as written will work correctly, but not due to the reasoning you have. From what I can tell the eager finalization logic executes during the time period where the background GC has the VM in a suspended state 100% of the time. This doesn't match the comments on the FinalizeWeakReference function, but should work. @Maoni0 could you tell me if I've missed something here?

Reply:
@davidwrighton You are right. @Maoni0 told me the same. 

As it is today, the runtime is suspended after the call to `bgc_suspend_ee` on `background_mark_phase` and is not resumed until `restart_EE` in `background_sweep`.

I was confused because @maoni0 also mention we _will_ make it concurrent, and in that case, the comment on `FinalizeWeakReference` _will_ be correct, and we _will_ need the `allow_fgc` mechanism.

Sorry for the confusion.

Reply:
Ok, please add a comment that the suspension state of the runtime must be prevented from changing while in this function. Currently, we are safe, and when we add the concurrent logic that Maoni suggests, we'll have to evaluate this again.